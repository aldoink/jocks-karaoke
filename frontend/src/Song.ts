export class Song {
    constructor(
        public location: string,
        public title: string,
        public artist: string) {}
}