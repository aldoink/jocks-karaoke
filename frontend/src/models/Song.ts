export class Song {
    constructor(
        public id: number,
        public location: string,
        public title: string,
        public artist: string
    ) {}
}