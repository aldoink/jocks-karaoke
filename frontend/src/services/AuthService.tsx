import axios from "axios";

export const BACKEND_URL = process.env.REACT_APP_BACKEND_URL;

export class AuthService {
    async login (email: string, password: string): Promise<void> {
        const response = await axios.post(`${BACKEND_URL}/auth/login`, {email, password});
        localStorage.setItem("token", response?.data?.token);
    }

    getToken() {
        return localStorage.getItem("token")
    }
}