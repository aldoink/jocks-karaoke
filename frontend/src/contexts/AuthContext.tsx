import {createContext} from "react";
import {AuthService} from "../services/AuthService";

export const AuthContext = createContext({authService: {} as AuthService});