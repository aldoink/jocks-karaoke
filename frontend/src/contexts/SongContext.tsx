import { createContext, Dispatch, SetStateAction } from "react";
import { SongService } from "../services/SongService";
import { Song } from "../models/Song";

export const songService = new SongService();
export interface ISongContext {
  readonly songService: SongService;
  songs: Song[];
  setSongs: Dispatch<SetStateAction<Song[]>>;
}

export const SongContext = createContext<ISongContext>({
  songService,
  songs: [],
  setSongs: () => {},
});
