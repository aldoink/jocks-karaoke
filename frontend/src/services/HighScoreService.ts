import axios, {AxiosRequestConfig} from "axios";
import {AuthService, BACKEND_URL} from "./AuthService";

export interface HighScore {
    [index: string]: string | number | undefined;

    name?: string;
    score?: string;
    songId: number;
}

export class HighScoreService {
    constructor(private authService: AuthService) {
    }

    public async findAll(songId: number): Promise<HighScore[]> {
        const response = await axios.get(`${BACKEND_URL}/highscores/${songId}`, this.getAxiosConfig());
        return response?.data;
    }

    public async save(entry: HighScore) {
        return await axios.put(
            `${BACKEND_URL}/highscores`,
            entry,
            this.getAxiosConfig()
        )
    }

    private getAxiosConfig(): AxiosRequestConfig {
        return {headers: {Authorization: "Bearer " + this.authService.getToken()}};
    }
}