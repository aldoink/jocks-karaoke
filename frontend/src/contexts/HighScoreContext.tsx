import { createContext } from "react";
import { HighScore, HighScoreService } from "../services/HighScoreService";
import { authService } from "./AuthContext";
import { Song } from "../models/Song";

export const highScoreService = new HighScoreService(authService);

export interface IHighScoreContext {
  highScoreService: HighScoreService;
  highScores: HighScore[];
  refreshHighScores: (song: Song) => void;
  hasError: boolean;
}

export const HighScoreContext = createContext<IHighScoreContext>({
  highScoreService,
  highScores: [],
  refreshHighScores: () => {},
  hasError: false,
});
