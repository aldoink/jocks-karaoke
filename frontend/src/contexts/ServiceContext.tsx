import {createContext} from "react";
import {HighScoreService} from "../services/HighScoreService";
import {SongService} from "../services/SongService";

export interface IServiceContext {
    readonly highScoreService: HighScoreService
    readonly songService: SongService
}

export const ServiceContext = createContext<IServiceContext>(
    {
        highScoreService: {} as HighScoreService,
        songService: {} as SongService
    }
);

