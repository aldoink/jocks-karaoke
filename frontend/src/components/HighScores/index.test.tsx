import "jest-styled-components";
import { HighScores } from "./index";
import { Song } from "../../models/Song";
import { render, screen } from "@testing-library/react";
import { HighScore, HighScoreService } from "../../services/HighScoreService";
import React from "react";
import { AuthService } from "../../services/AuthService";
import {
  HighScoreContext,
  IHighScoreContext,
} from "../../contexts/HighScoreContext";
import { AuthContext, IAuthContext } from "../../contexts/AuthContext";

describe("HighScoreList", () => {
  const defaultSong = new Song(1, "BP-0001", "Title", "Artist");
  const mockedHighScoreService = {
    findAll: jest.fn(),
  } as unknown as HighScoreService;
  const mockedAuthService = {} as AuthService;
  const refreshHighScores = jest.fn();
  const mockHighScoreList = [
    { name: "Ally", score: "99", songId: 1 } as HighScore,
    { name: "Jenny", score: "98", songId: 1 } as HighScore,
    { name: "Johnny", score: "97", songId: 1 } as HighScore,
    { name: "Gemma", score: "96", songId: 1 } as HighScore,
  ];

  function renderHighScoreList(contextValues?: Partial<IHighScoreContext>) {
    return render(
      <AuthContext.Provider
        value={{ authService: mockedAuthService } as IAuthContext}
      >
        <HighScoreContext.Provider
          value={
            {
              highScoreService: mockedHighScoreService,
              highScores: mockHighScoreList,
              refreshHighScores,
              hasError: false,
              ...contextValues,
            } as IHighScoreContext
          }
        >
          <HighScores song={defaultSong} />
        </HighScoreContext.Provider>
      </AuthContext.Provider>
    );
  }

  beforeEach(() => {
    mockedAuthService.isAuthenticated = jest.fn().mockReturnValue(true);
  });

  it("loads the high scores for the selected song", async () => {
    //given
    mockedHighScoreService.findAll = jest
      .fn()
      .mockResolvedValue(mockHighScoreList);

    const { asFragment } = renderHighScoreList();

    //then
    expect(refreshHighScores).toHaveBeenCalledWith(defaultSong);
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows an error message when high score context hasError", async () => {
    //given
    renderHighScoreList({ hasError: true });
    //then
    expect(
      await screen.findByText("Something went wrong... try again later.")
    ).toBeInTheDocument();
  });

  it("shows a message when no high scores are found", async () => {
    //given
    renderHighScoreList({ highScores: [] });

    //then
    expect(
      await screen.findByText(
        "Doesn't look like anyone's set a high score yet!"
      )
    ).toBeInTheDocument();
  });

  it('shows an "Add High Score" button when the user is authenticated', () => {
    //given
    mockedAuthService.isAuthenticated = jest.fn().mockReturnValue(true);
    renderHighScoreList();

    //then
    expect(screen.getByText("Add new High Score")).toBeInTheDocument();
  });

  it('does not show an "Add High Score" button when the user is not authenticated', () => {
    //given
    mockedAuthService.isAuthenticated = jest.fn().mockReturnValue(false);
    renderHighScoreList();

    //then
    expect(screen.queryByText("Add new High Score")).not.toBeInTheDocument();
  });

  it('hides the "Add new High Score" button when addMode is enabled', () => {
    //given
    mockedAuthService.isAuthenticated = jest.fn().mockReturnValue(true);
    renderHighScoreList();

    ///when
    screen.getByText("Add new High Score").click();

    //then
    expect(screen.queryByText("Add new High Score")).not.toBeInTheDocument();
  });

  it("shows the AddHighScore component when the Add High Score button is clicked", () => {
    //given
    mockedAuthService.isAuthenticated = jest.fn().mockReturnValue(true);
    renderHighScoreList();

    //when
    screen.getByText("Add new High Score").click();

    //then
    expect(screen.getByPlaceholderText("Name")).toBeInTheDocument();
    expect(screen.getByPlaceholderText("Score")).toBeInTheDocument();
  });

  it('does not show the AddHighScore component before the user clicks on the "Add new High Score" button', () => {
    //given
    mockedAuthService.isAuthenticated = jest.fn().mockReturnValue(true);
    renderHighScoreList();

    //then
    expect(screen.queryByPlaceholderText("Name")).not.toBeInTheDocument();
    expect(screen.queryByPlaceholderText("Score")).not.toBeInTheDocument();
  });

  it("hides the AddHighScore component when the user clicks on the 'Cancel' button", () => {
    //given
    mockedAuthService.isAuthenticated = jest.fn().mockReturnValue(true);
    renderHighScoreList();

    //when
    screen.getByText("Add new High Score").click();
    screen.getByText("Cancel").click();

    //then
    expect(screen.queryByPlaceholderText("Name")).not.toBeInTheDocument();
    expect(screen.queryByPlaceholderText("Score")).not.toBeInTheDocument();
  });
});
