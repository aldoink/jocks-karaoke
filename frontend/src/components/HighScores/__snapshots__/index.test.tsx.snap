// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighScoreList loads the high scores for the selected song 1`] = `
<DocumentFragment>
  .c1 {
  width: 100%;
  border-collapse: collapse;
}

.c2 td:first-child {
  width: 75%;
  margin-right: 1rem;
}

.c2 td CenteredColumn {
  width: 25%;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c5 {
  margin-top: 1rem;
  padding: 0.25rem 0.75rem;
  border: 1px solid;
  border-radius: 4px;
  background-color: transparent;
  font-size: 18px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 {
  border-bottom: 1px solid red;
}

.c0 h2,
.c0 h3 {
  margin: 0;
}

.c0 h4 {
  margin-top: 8px;
}

<div
    class=""
  >
    <div
      class="c0"
    >
      <h2>
        Title
      </h2>
      <h3>
        Artist
      </h3>
      <h4>
        BP-0001
      </h4>
    </div>
    <div>
      <table
        class="c1"
      >
        <thead>
          <tr>
            <th
              class="name-column"
            >
              Name
            </th>
            <th
              class="score-column"
            >
              Score
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="c2"
          >
            <td>
              Ally
            </td>
            <td
              class="c3"
            >
              99
            </td>
          </tr>
          <tr
            class="c2"
          >
            <td>
              Jenny
            </td>
            <td
              class="c3"
            >
              98
            </td>
          </tr>
          <tr
            class="c2"
          >
            <td>
              Johnny
            </td>
            <td
              class="c3"
            >
              97
            </td>
          </tr>
          <tr
            class="c2"
          >
            <td>
              Gemma
            </td>
            <td
              class="c3"
            >
              96
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="c4"
      >
        <button
          class="c5"
        >
          Add new High Score
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
